### DISCUSSION TITLE
Code-Only / Minimal Stride API Project

### DISCUSSION THREAD

**Discussion by VaclavElias** [stride3d/stride#1253] [Ideas]
I have been working with ASP.NET Core intensively and was happy when they added Minimal API to simplify the setup and learning process, this got even better with .NET 6 which allows the use of **global usings** and **file-scoped namespaces**. Please be aware I am not a game developer but I like to experiment 🤣🙏.

_Note: In this conversation **API** means **Application Programming Interface**._

> Minimal APIs are architected to create APIs with minimal dependencies. They are ideal for apps that want to include only the minimum files, features and dependencies.

There might be other reasons, why not to use a game editor, including:

- You don't want to install anything on your computer (no Stride installation required?)
- You want to start very quickly
- You want to learn C# programming with a nice visual output instead of console
- You want to learn game programming **gradually, in the simplest way**, without using the game editor
- You find coding and coding tools very complex to understand and navigate around
- You want to start with game development basics before you even start exploring the game editor
- Easy and quick prototyping
- Easy to learn game development concepts and steps

I got inspired by multiple projects here and there from GitHub, some issues (#80, #986), our Discord discussions and help from the community and I created a working prototype for us. Ideally it should be very flexible, starting very simple but an experienced dev can still do what they want to do, and maybe eventually it could be also a hybrid and easy to hook to Stride Studio Editor to add anything else from the Editor.

https://github.com/VaclavElias/Stride3DTutorials/tree/main/Other/Minimal

_Note: Any name convention can be changed as I get your feedback._

- [x] Initial discussions - this was discussed on Discord channel and also #80, #986
- [x] Preview Concepts - a few concepts will be proposed
- [X] Issue Opened - inviting others to suggest features and review - https://github.com/stride3d/stride/issues/1295
- [x] Naming Convention - finalising naming convention for methods, properties, namespaces
- [x] File locations
- [x] Tests
- [x] Beta Release

# Review 1 - Concept

The main part is a new namespace **Stride.Engine.Builder**, which is going to be a wrapper for a boilerplate code. My prototype doesn't need to make any changes in the Game Engine but maybe some changes could make my implementation easier by exposing new event handlers or whatever you think would be better.

**One NuGet package added to a new console application will make Stride to shine just with 21 lines or less** 🌟🌟🌟

_Note: The Add methods can't be before the .Build() because they need Game object but we can figure out a better logic here._

```csharp
var builder = GameApplication.CreateBuilder();

var game = builder.Build();

builder.AddGround();
builder.AddSkybox();
builder.AddCameraController();

builder.Add(GetCubeEntity(),new CubeProceduralModel());
builder.Add(new SphereProceduralModel());

game.Run();

Entity GetCubeEntity()
{
    var entity = new Entity();

    entity.Transform.Position = new Vector3(1,0,3);

    return entity;
}
```

I have got these questions for us:

1. Can we make my implementation even simpler, cleaner and elegant?
2. Can we make GameApplication any simpler?
3. What changes would you suggest?
4. Could we add this as a new optional official Stride NuGet package once finalised, as an -alpha package?
5. I placed some of my logic to `BeginRun()` in the **MinimumGame** class as I found that the only place this works but I don't like that I had to create a **MinimalGame** class. If we had a new EventHandler also for BeginRun (or wherever it is better) like we have Activated, Deactivated, Exiting, WindowCreated handlers in the **GameBase**, then I don't need MinimalGame class and would do the full set up in the **GameApplication.cs**. Do we need to update Stride **GameBase** class to simplify my implementation?
6. <del>Can anyone help/suggest how to improve the visual output, am I missing just material and lighting?
  -- The default could be as it is (first image)
  -- And improved should look like on the second image (done in the editor as an example) or better, so anyone working with this NuGet package would be very impressed by Stride engine</del> ..resolved
7. <del>What else could we add as optional? e.g. like AddSkybox, better lightning and so on.</del> ..resolved

Generated by this code-only project
![image](https://user-images.githubusercontent.com/4528464/147798782-dbd37948-4c1f-44a2-9e5a-b4bf598c5505.png)

Please feel free to add any comments or questions.

Thank you!

**Comment by RelentlessArts** [comment]
Would this allow the building of a headless stride server? If so I am very interested in this.

**Reply by VaclavElias** [reply]
To be honest I have no idea it if would be possible and I am not sure how far from this it would be. Maybe @Aggror or @tebjan or @manio143 might answer your question here.

**Reply by manio143** [reply]
The current issue of headless Stride is that there's things related to input/graphics/audio integrated in the `Game` and `ScriptComponent`, so a purely logic based server would need to reimplement some of those systems (or they would need to be redesigned in the engine for higher composability). There was work done @herocrab towards it but it's generally not a simple task. You can see some discussions around it on Discord.

**Reply by RelentlessArts** [reply]
Thank you @VaclavElias &  @manio143  I've seen a few of the discussions before so I'll have another look, I was just a bit hopeful that this could provide a stripped down headless version of stride.

**Reply by najak3d** [reply]
@VaclavElias  - WOW!   This is a great selling point for Stride, truly!    I am very impressed with the ease of using Stride without it's editor. 

Here are the real-world applications I have in mind:
1. Application simply wants a new renderer!...    Currently we're using Urho.NET (wrapped Urho3D), and we are looking for possible replacement, which means we're just wanting the rendering... not even physics, and ALL content is programmatically created.

2. MAUI + Stride3D -  This plugin needs to be written, and will want this Simple API to use -- again for non-game apps that simply want some 3D presentations.    This is a suped-up replacement for the basic OpenGL renderer.    Right now, MAUI doesn't have a "real" game engine integrated with it.   I think it should be Stride.

3. Integration with Avalonia -- same deal, same concept as #.2.


Just looking at the Stride references you made, this is only 5.5 MB of DLL's.   That's slim, and can even be smaller if you exclude Audio/Physics/VR/etc.


The vital thing here, to make Stride *really* take off, IMO, is to fill this .NET void for "3D Renderers" that fit (and retrofit) into more application spaces.

Urho.NET even allows Avalonia to render into it's own UI.... so we now have Avalonia UI INSIDE Urho.NET.   It's not a tough concept to implement, and could be done for Stride as well. 

.NET MAUI also has a "DrawnControls" project, which delivers pixel-perfect controls, renderable inside of a .NET Game window (as an image that is updated by Maui logic).    Just as we've done with Avalonia + Urho.NET, we could do this with MAUI + Stride.

How sweet of an offering would it be to have Stride3D offer:
1. Renders inside a MAUI control (e.g. picture box), simply (no editor).
2. Allows you to have MAUI "DrawControls" rendered inside a Stride Game.
3. Same as #.1 or #.2 but with Avalonia.

Now an app written in either MAUI or Avalonia, can easily choose to use Stride3D for it's 2D/3D graphics presentation needs -- and can even use their UI of choice INSIDE Stride (Avalonia or MAUI).

What I am describing here is what I believe to be the long-standing VOID for .NET developers, trying to find a good solution for 2D/3D graphics presentation.

Our team is planning on pushing forward ONE solution -- it's either going to be:
1. Urho.NET
2. Unity3D
3. Stride.


Unity3D is appealing, but it's commercialized.    And so it won't be open-source.   Also it's C++, so when things don't work as you think they should -- you can't really debug it!    However, Unity, by far, has the most momentum and biggest community with advanced plugins/techniques.

Urho.NET is open-source, but is lacking momentum.   Also it's C++ -- so when things don't work, debugging is an issue (at least here you can see their source code in C++ and might be able to discern what went wrong, even if you can't live debug it).

Stride3D - is the only pure C#/.NET solution.   Which might make it harder for AA games, but still should work just fine for most Indie games not requiring top-performance for large worlds/etc.   I'm not sure about the community here, but seems fairly strong.   Once it's integrated to MAUI + Avalonia -- the momentum will build even more.

**Reply by VaclavElias** [reply]
@najak3d thanks for your feedback here. It is great to know that there are other possibilities/advantages of using code-only approach. I am taking it only to the first step as I am not that experienced in game development so any further input and help from others will be welcome to take it to the next level.

**Comment by VaclavElias** [comment]
# Review 2

I have pushed a new version where there are now 2 approaches. I am going to use/test this now for a few weeks and see how this works.

**Namespace Stride.Engine.Builder**
- Creates default Scene, the original **SceneBaseFactory** can't be used because it is using Assets which are currently bind to Windows, some code was used
- **BasicCameraController.cs** was used from some examples, it would be better if there is official and maintained one and directly in the Engine 
- **GraphicsCompositorHelper**  to create a default **GraphicsCompositor** could be used directly from the Game Engine, however I needed to update it
- the original **SkyboxGenerator.cs** can't be used because it is using Assets which are currently bind to Windows, some code was used

**New Approach which I like more**

- It is using _EventHandler_
- It could be added to Game Engine (2 lines), so we don't have to create this _MinimalGame_ class at all

```C#
public class MinimalGame : Game
{
    public event EventHandler<EventArgs>? OnBeginRun;

    protected override void BeginRun()
    {
        base.BeginRun();

        Window.AllowUserResizing = true;

        OnBeginRun?.Invoke(this, EventArgs.Empty);
    }
}
```

- _Build3D()_ adds DefaultMaterial, Ground, Skybox, CameraScript, Sphere
- GetCube() is just a wrapper to Generate Cube (it just hides the complexity..)

![image](https://user-images.githubusercontent.com/4528464/147836644-c072f50c-86e7-4d54-ae5c-7749cfa82033.png)

**The other (previous) approach**
- It is using _Action_

```C#
public class MinimalGame : Game
{
    public List<Action> BeginRunActions { get; set; } = new();

    protected override void BeginRun()
    {
        base.BeginRun();

        Window.AllowUserResizing = true;

        foreach (var action in BeginRunActions)
        {
            action.Invoke();
        }
    }

    public void AddAction(Action? action)
    {
        if (action == null) return;

        BeginRunActions.Add(action);
    }
}
```
Same as in the first case but uses _AddAction_
![image](https://user-images.githubusercontent.com/4528464/147836938-2473eba1-05ce-42da-b340-3488799be3ec.png)

![image](https://user-images.githubusercontent.com/4528464/147837413-583276d8-7862-45f8-831e-242ca70aeeb1.png)

Let me know your comments :)

**Reply by Eideren** [reply]
Nice job already @VaclavElias !

Right now our setup to run a game is to do
```cs
using (var game = new Game())
{
    game.Run();
}
```
This is so much simpler to find and setup than yours right now, but yes, ours has a lot of issues that you dealt with, so would it be possible to use this pattern and layer *all* your fixes on top of it by creating a new game type deriving from ``Game`` ?
If that's the case we may as well make your changes part of the actual engine implementation of ``Game`` so that users benefit from this directly, won't need to explicitly find a minimal API and wouldn't have to work to 'upgrade' from the would be minimal API either ?
Something that could look like this:
```cs
using (var game = new Game{ /*Can change properties here*/ })
{
    game.Run( /*Can specify requirements here*/, /*those two would be optional parameters*/start: Start, update: Update );

    Entity entity = null;
    void Start(Scene rootScene, ServiceRegistry services)
    {
        entity = new Entity(new Vector3(1f, 0.5f, 3f))
        {
            new ModelComponent(new CubeProceduralModel().Generate(services)),// needs overload which creates a model and returns it
        };
        entity.Scene = rootScene;
    }
    void Update(SceneInstance rootScene, ServiceRegistry services, GameTime time)
    {
        entity.Transform.Position = new Vector3((float)Math.Sin(time.Elapsed.TotalSeconds), 0f, 0f);
    }
}
```
Or like this to ease the access to refs generated inside start from update:
```cs
using (var game = new Game{ /*Can change properties here*/ })
{
    game.Run( /*Can specify requirements here*/, start: Start );

    void Start(Scene rootScene, ServiceRegistry services, out Action<GameTime> update)
    {
        Entity entity = new Entity(new Vector3(1f, 0.5f, 3f))
        {
            new ModelComponent(new CubeProceduralModel().Generate(services)),
        };
        entity.Scene = rootScene;
        update = time =>
        {
            entity.Transform.Position = new Vector3((float) Math.Sin(time.Elapsed.TotalSeconds), 0f, 0f);
        };
    }
}
```
But that would be perhaps too obtuse to use or reason about for users ?
I haven't looked at your builder so not sure how trivial that would be to adapt, what do you think ?

**Reply by VaclavElias** [reply]
Thanks for the feedback @Eideren, I like your ideas even better, let me explore and play with it, I will reimplement my bits and come back here once I am done.

**Comment by VaclavElias** [comment]
# Review 3

https://github.com/VaclavElias/Stride3DTutorials/tree/main/Other/Minimal

## Option 1 - Update in Game.cs

I made another working prototype with some above suggestions from @Eideren implemented. Any suggested naming convention can be changed of course.

The reason that I didn't pass **GameDefaults** to Game/Run is because it can be actually separated in this prototype. You need to look into GameDefaults what it does and see if it makes sense to pass it to Game or Run, or keep it separately as it is.

```C#
using (var game = new Game())
{
    var _entity = new Entity(new Vector3(1f, 0.5f, 3f));
    var _angle = 0f;
    var initialPosition = _entity.Transform.Position;

    // override for game.Run()
    // start and update is optional, you can add any Update() through traditional script _entity.Components.Add(MyScript);
    // it is using default engine Script Start and Update
    game.Run(start: Start, update: Update);

    void Start()
    {
        // adds default GraphicsCompositor, Scene, Skybox, CameraScript, Ground, Material
        var defaults = new GameDefaults(game).Set3D();

        // or pick up what you want
        var defaults = new GameDefaults(game).Set().AddGround().AddSkybox().AddCameraScript().AddGameProfiler();

        var model = new CubeProceduralModel().Generate(game.Services);
        
        // I am getting DefaultMaterial here
        model.Materials.Add(defaults.DefaultMaterial);

        _entity.Components.Add(new ModelComponent(model));

        _entity.Scene = game.SceneSystem.SceneInstance.RootScene;
    }

    void Update()
    {
        _angle += 5f * (float)game.UpdateTime.Elapsed.TotalSeconds;

        var offset = new Vector3((float)Math.Sin(_angle), 0, (float)Math.Cos(_angle)) * 1f;

        _entity.Transform.Position = initialPosition + offset;
    }
}
```
## Engine Game.cs
Around 14 lines of code added

```c#
public class Game : GameBase, ISceneRendererContext, IGameSettingsService
{
   private RootScript rootScript;

   protected override void Initialize()
   {
      ...
      // Add root script if exists
      if (rootScript != null)
      {
         Script.Add(rootScript);
      }
      ... 
   }

   // this runs before Initialize() and we can't use Script.Add() here because Script doesn't exist yet
   public void Run(Action start, Action update)
   {
       if (start != null || update != null)
       {
           rootScript = new RootScript(start, update);
       }

       base.Run();
   }

}
```

New class RootScript.cs 

```c#
public class RootScript : SyncScript
    {
        private readonly Action? startAction;
        private readonly Action? updateAction;

        public RootScript(Action? start = null, Action? update = null)
        {
            startAction = start;
            updateAction = update;
        }

        public override void Start()
        {
            startAction?.Invoke();
        }

        public override void Update()
        {
            updateAction?.Invoke();
        }
    }
```
## Option 2 - Update in GameBase.cs

Depends where **GameDefaults** is (before Run() or inside Start()) affects the GameDefaults implementation.

```c#
using (var game = new Game())
{
    // adds default camera, camera script, skybox, ground, ..like through UI
    var defaults = new GameDefaults(game).Set3D();

    // only Start, which is called once the game is running
    game.Run(start: Start);

    void Start()
    {
        var model = new CubeProceduralModel().Generate(game.Services);

        model.Materials.Add(defaults.DefaultMaterial);
        
        // if you need Update(), just add your custom script, I added MotionComponentScript
        var entity = new Entity(new Vector3(1f, 0.5f, -3f))
        {
            new ModelComponent(model),
            new MotionComponentScript()
        };

        entity.Scene = game.SceneSystem.SceneInstance.RootScene;
    }
}
```

## Engine GameBase.cs
Around 4 lines of code added

```c#

    public abstract class GameBase : ComponentBase, IGame
    {
       ...
      // maybe it can be private 
      public Action? StartAction;
       ...

        // "Action? start = null" added
        public void Run(GameContext gameContext = null, Action? start = null)
        {
            ...
            StartAction = start;
           ...
        }

        protected virtual void BeginRun()
        {
            // only this line added
            StartAction?.Invoke();
        }

```

Let me know what you think.

**Reply by Eideren** [reply]
Sorry I couldn't get to this sooner !
I think option one would be more appropriate as I think ``GameBase`` should only contain the bare minimum implementation of a game.

Now I had a couple of thoughts about your ``GameDefaults``, how about we change this into a static class containing extension methods around the ``Game`` type, something like this:
```cs
public static class GameDefaults
{
   public static Entity AddGround(this Game game)
   {
      // Do it right away instead of buffered
      var materialDescription = new MaterialDescriptor
      {
         Attributes =
            {
               Diffuse = new MaterialDiffuseMapFeature(new ComputeColor(Color.FromBgra(0xFF242424))),
               DiffuseModel = new MaterialDiffuseLambertModelFeature(),
      // Rest of the implementation ...


      return groundEntity; // So that caller can move, rotate it or do other modification if need be
   }
}
```
At the top of your file:
```cs
using static Stride.GameDefaults;
```
Then implementation would look like so:
```cs
using (var game = new Game())
{
   var _entity = new Entity(new Vector3(1f, 0.5f, 3f));
   var _angle = 0f;
   var initialPosition = _entity.Transform.Position;

   game.Run(start: Start, update: Update);

   void Start(Scene rootScene, ServiceRegistry services)
   {
      game.SetupBase3DScene();
      // or pick up what you want
      game.AddSkybox();
      var groundEntity = game.AddGround();
      game.AddMouseLookCamera();


      var model = new CubeProceduralModel().Generate(services);
      
      model.Materials.Add(game.NewDefaultMaterial());

      _entity.Components.Add(new ModelComponent(model));

      _entity.Scene = rootScene;
   }

   void Update(SceneInstance rootScene, ServiceRegistry services, GameTime time)
   {
      _angle += 5f * (float)time.Elapsed.TotalSeconds;
      var offset = new Vector3((float)Math.Sin(_angle), 0, (float)Math.Cos(_angle));
      _entity.Transform.Position = initialPosition + offset;
   }
}
```
To me it looks less painful to maintain, update and work with, no need to make sure you created an implementation, you can use it from anywhere even in ``SyncScript`` and the likes as you're not hooking specifically into ``OnBeginRun`` ?
This is quite a big departure though so it might be too far of ? What do you think ?

**Reply by VaclavElias** [reply]
Let me digest your comments but I like them very much 😀. No need for sorry, I also need to find some spare slots for this 🤣.

**Reply by manio143** [reply]
Personally I'm not a huge fan of adding `RootScript` directly as a field to `Game`. It'd be better to provide more events in different places to hook into. This way instead of creating an overload of `Run` you would create an extension method like

```csharp
public static void Run(this Game game, Action start, Action update)
{
    if (start != null || update != null)
    {
        var rootScript = new RootScript(start, update);
        game.OnInitialize += () => game.Script.Add(rootScript);
    }

    game.Run();
}
```

This way someone else can come along and add their own logic to `OnInitialize`.

**Reply by VaclavElias** [reply]
Thanks @manio143, I like that as well, let me test it and I will update you.

**Comment by VaclavElias** [comment]
# Review 4
Feedback from @Eideren  and @manio143.

Thanks for your feedback. I think your feedback can be merged together. Note again, that the naming convention can be changed to anything appropriate at the later stage.

If I forgot to address anything you mentioned previously, please feel free to mention that again. Once our thoughts are more stable. I will ping all others who expressed interest in the code-only approach.

Working prototypes:
- https://github.com/VaclavElias/Stride3DTutorials/tree/main/Other/Minimal
- https://github.com/VaclavElias/stride/tree/code-only

Regarding the method signature which I left below as an example I have got these findings:

- I wouldn't add there `IServiceRegistry`, because we can easily reach it with `game.Services`
- **GameTime** is probably ok, because it is deeper, unless we would like to take it further and we would use TimeSpan which would return Elapsed
- With **Scene**, there are a few issues
  - The Scene is created through `game.SetupBase3DScene()`, so it doesn't make sense to pass it through signature because it doesn't exist yet.
  - It could be passed if 
    - The Scene was created in the RootScript
    - Or there would be another extension `game.SetupBase()` called before `game.Run(start: Start, update: Update)` in the user project
    - Or by updating **SceneSystem.cs** `LoadContent()` and adding the below here https://github.com/stride3d/stride/blob/b9b2dbb380f39ea02dab3c65cc0ae1b68e7c2f51/sources/engine/Stride.Engine/Engine/SceneSystem.cs#L122
    ```c# 
    if (sceneTask == null)
    {
        SceneInstance ??= new SceneInstance(Services) { RootScene = new Scene() };
    }
    ```
    - Or we don't have to have this convenience yet, maybe to start with, I wouldn't add it and based on the further feedback from others we could elaborate on this further. Unless you can see from your experience that it is worth it to implement it.

## Engine Game.cs
2 lines of code added instead of 14.

```c#
public Action OnInitialize;  //new line, not sure if you want to have it this way, using Action 

protected override void Initialize() 
{ 
  ... 
  OnInitialize?.Invoke(); //new line, end of the method. 
}
```

## RootScript.cs
Based on your feedback, it might be simplified just to
```c#
public class RootScript : SyncScript
{
    private readonly Action? startAction;
    private readonly Action<GameTime>? updateAction;
    
    public RootScript(Action? start = null, Action<GameTime>? update = null)
    {
        startAction = start;
        updateAction = update;
    }

    public override void Start()
    {
        startAction?.Invoke();
    }

    public override void Update()
    {
        updateAction?.Invoke(Game.UpdateTime);
    }    
}    
```
If we want other bits and we figure out how to get Scene
```c#
public class RootScript : SyncScript
{
    private readonly Action<Scene, IServiceRegistry>? startAction;
    private readonly Action<Scene, IServiceRegistry, GameTime>? updateAction;

    public RootScript(Action<Scene, IServiceRegistry>? start = null, Action<Scene, IServiceRegistry, GameTime>? update = null)
    {
        startAction = start;
        updateAction = update;
    }

    public override void Start()
    {
        //create empty scene here so it can be passed
        //SceneSystem.SceneInstance ??= new SceneInstance(Services) { RootScene = new Scene() };

        startAction?.Invoke(SceneSystem.SceneInstance?.RootScene, Services);
    }

    public override void Update()
    {
        updateAction?.Invoke(SceneSystem.SceneInstance?.RootScene, Services, Game.UpdateTime);
    }
}
```
## GameExtensions.cs
https://github.com/VaclavElias/Stride3DTutorials/blob/main/Other/Minimal/Stride.GameDefaults/GameExtensions.cs

## Code Only Examples
### Example 1 .NET 6, C# 10
Nice and clean, encouraging using Components approach and using Script Components.
```c#
using var game = new Game();

game.Run(start: Start, update: null);

void Start(Scene rootScene, IServiceRegistry services)
{
    // adds scene, graphics comositor, camera, camera script, skybox, ground ..like through UI
    game.SetupBase3DScene();

    var entity = new Entity(new Vector3(1f, 0.5f, 3f))
    {
        new ModelComponent(new CubeProceduralModel().Generate(services)),
        new MotionComponentScript()
    };

    entity.Scene = game.SceneSystem.SceneInstance.RootScene;
}
```
or
```c#
using var game = new Game();

game.Run(start: Start, update: null);

void Start(Scene rootScene, IServiceRegistry services)
{
    // Cherry Pick
    game.SetupBase(); // required, adds scene, graphics comositor, camera
    //game.AddSkybox();
    //game.AddGround();
    game.AddMouseLookCamera();
    game.AddProfiler();

    var entity = new Entity(new Vector3(1f, 0.5f, 3f))
    {
        new ModelComponent(new CubeProceduralModel().Generate(services)),
        new MotionComponentScript()
    };

    entity.Scene = game.SceneSystem.SceneInstance.RootScene;
}
```

### Example 2
Using ```Update()```, ```game.NewDefaultMaterial()``` extension.
```c#
using (var game = new Game())
{
    var entity = new Entity(new Vector3(1f, 0.5f, 3f));
    var angle = 0f;
    var initialPosition = entity.Transform.Position;

    game.OnInitialize += () => System.Console.WriteLine("Hello, manio143");

    game.Run(start: Start, update: Update);

    void Start(Scene rootScene, IServiceRegistry services)
    {
        game.SetupBase3DScene();

        var model = new CubeProceduralModel().Generate(services);

        model.Materials.Add(game.NewDefaultMaterial());

        entity.Components.Add(new ModelComponent(model));

        entity.Scene = game.SceneSystem.SceneInstance.RootScene;
    }

    void Update(Scene rootScene, IServiceRegistry services, GameTime time)
    {
        angle += 1f * (float)time.Elapsed.TotalSeconds;

        var offset = new Vector3((float)Math.Sin(_angle), 0, (float)Math.Cos(_angle)) * 1f;

        entity.Transform.Position = initialPosition + offset;
    }
}
```

Let me know what you think 😀.

**Reply by Eideren** [reply]
I think it would make sense for the `SceneInstance` to create a default empty scene if nothing has been set up, so like you said here:
https://github.com/stride3d/stride/blob/b9b2dbb380f39ea02dab3c65cc0ae1b68e7c2f51/sources/engine/Stride.Engine/Engine/SceneSystem.cs#L115-L121
I would add an else statement below it:
```cs
}
else
{
    SceneInstance ??= new SceneInstance(Services) { RootScene = new Scene() };
}
```
You're right for the service registry stuff, it's not useful enough to bother with including it as a parameter.
Here's an extension method I setup which works without any other changes engine-side

```cs
public static class GameExtension
{
    public static void Run(this Game game, GameContext? context = null, Action<Scene>? start = null, Action<Scene, GameTime>? update = null )
    {
        game.SceneSystem.SceneInstance ??= new SceneInstance(game.Services, new Scene()); // this would be removed once SceneInstance changes are merged in
        game.Script.Scheduler.Add(RootScript);
        game.Run(context);
        
        async Task RootScript()
        {
            start?.Invoke(GetRootScene());
            if(update == null)
                return;
            do
            {
                update.Invoke(GetRootScene(), game.UpdateTime);
                await game.Script.NextFrame();
            } while (true);
        }
        
        Scene GetRootScene() => game.SceneSystem.SceneInstance.RootScene;
    }
}
```
How's that ?

**Reply by VaclavElias** [reply]
And it does work 😀! Let me do a few more tests. So many ways to achieve this. I like that we don't touch actually anything. We lose ```OnInitialize``` which manio143 suggested but if there wasn't a demand for that then that's ok especially if we don't need it actually for this code-only approach.

**Reply by VaclavElias** [reply]
@Eideren I keep testing this in my examples and playing more. All seems ok. Wouldn't be this update a breaking change for some projects whoever was expecting a null (maybe very very few)? Also, any suggestion how could we test this? Probably, if we implement it then  if somebody would question functionality the test would fail if it was removed. I am not yet familiar with Stride tests so I am not sure if I can reuse any existing tests. 

If we go with this change, Shall I create a separate PR or I should keep this change together with other bits?

```c#
            else
            {
                SceneInstance ??= new SceneInstance(Services) { RootScene = new Scene() };
            }
```

**Reply by Eideren** [reply]
@VaclavElias I doubt anyone would have an issue with this, no one to my knowledge is doing code only right now and I doubt of those people they would have a null check on this instead of just assigning it straight away. There's nothing setup to test that class yet and I don't think doing test for this specific thing is worth it.
Yes, please do a separate PR for this, thanks !

**Reply by VaclavElias** [reply]
In that case I am going to move to the next steps as our brainstorming and concepts are finalised here.

- I will do PR for the SceneInstance
- I am going to do some clean up code and further testing my examples
- I will create an Issue for Code-Only Implementation so we could get more feedback and wrap it up?

Other Steps listed on the top

Is that ok?

**Reply by Eideren** [reply]
Sure

**Reply by najak3d** [reply]
It all looks good to me.   My only concern is to know that you are going to enable the inclusion of many samples into a single SLN, so that you build/run-once, and from there can browse through the various feature samples.

**Reply by VaclavElias** [reply]
@najak3d I think it should be ok when I start working on these examples, as you suggested it can be all in one SLN and have some UI to list and launch individual examples.

**Reply by Aggror** [reply]
Excellent work on this @VaclavElias 

We could eventually extend the documentation website, so that it references these actual code files. The C# beginner tutorials do this too.

**Comment by VaclavElias** [comment]
Issue opened here https://github.com/stride3d/stride/issues/1295.

**Comment by VaclavElias** [comment]
Just adding a note. The Code Only is now supported through the Stride Community Toolkit, where all discussed above was implemented https://stride3d.github.io/stride-community-toolkit/manual/code-only/index.html

**Reply by tebjan** [reply]
Thanks, as always, great work!
